enaml/src/subscription_observer.cpp

```diff
@@ -0,0 +1,332 @@
+/*-----------------------------------------------------------------------------
+ * | Copyright (c) 2025, Nucleic Development Team.
+ * |
+ * | Distributed under the terms of the Modified BSD License.
+ * |
+ * | The full license is in the file LICENSE, distributed with this software.
+ * |----------------------------------------------------------------------------*/
+#include <iostream>
+#include <sstream>
+#include <cppy/cppy.h>
+
+#ifdef __clang__
+#pragma clang diagnostic ignored "-Wdeprecated-writable-strings"
+#endif
+
+#ifdef __GNUC__
+#pragma GCC diagnostic ignored "-Wwrite-strings"
+#endif
+
+
+namespace enaml
+{
+
+// ptr to atom.api.atomref
+static cppy::ptr atomref;
+
+// POD struct - all member fields are considered private
+struct SubscriptionObserver
+{
+    PyObject_HEAD
+    PyObject* atomref;
+    PyObject* name;
+
+    static PyType_Spec TypeObject_Spec;
+
+    static PyTypeObject* TypeObject;
+
+    static bool Ready();
+
+    static bool TypeCheck( PyObject* ob );
+
+};
+
+
+namespace
+{
+
+
+PyObject*
+SubscriptionObserver_new( PyTypeObject* type, PyObject* args, PyObject* kwargs )
+{
+    PyObject* owner;
+    PyObject* name;
+    static char* kwlist[] = { "owner", "name", 0 };
+    if( !PyArg_ParseTupleAndKeywords( args, kwargs, "OU", kwlist, &owner, &name ) )
+    {
+        return 0;
+    }
+    cppy::ptr ptr( PyType_GenericNew( type, args, kwargs ) );
+    if( !ptr )
+    {
+        return 0;
+    }
+    SubscriptionObserver* self = reinterpret_cast<SubscriptionObserver*>( ptr.get() );
+
+    self->atomref = PyObject_CallOneArg(atomref.get(), owner);
+    if( !self->atomref )
+    {
+        return 0;
+    }
+    self->name =  cppy::incref( name );
+    return ptr.release();
+}
+
+
+void
+SubscriptionObserver_clear( SubscriptionObserver* self )
+{
+    Py_CLEAR( self->atomref );
+    Py_CLEAR( self->name );
+}
+
+
+int
+SubscriptionObserver_traverse( SubscriptionObserver* self, visitproc visit, void* arg )
+{
+    Py_VISIT( self->atomref );
+    Py_VISIT( self->name );
+    Py_VISIT(Py_TYPE(self));
+    return 0;
+}
+
+
+void
+SubscriptionObserver_dealloc( SubscriptionObserver* self )
+{
+    PyObject_GC_UnTrack( self );
+    SubscriptionObserver_clear( self );
+    Py_TYPE(self)->tp_free( reinterpret_cast<PyObject*>( self ) );
+}
+
+
+int
+SubscriptionObserver__bool__( SubscriptionObserver* self )
+{
+    return PyObject_IsTrue( self->atomref );
+}
+
+
+/*
+ * Calls engine update with the owner and name
+ * if self.ref:
+ *     owner = self.ref()
+ *     engine = owner._d_engine
+ *      if engine is not None:
+ *         engine.update(owner, self.name)
+ */
+PyObject*
+SubscriptionObserver_call( SubscriptionObserver* self, PyObject* args, PyObject* kwargs )
+{
+
+    if( PyObject_IsTrue( self->atomref ) )
+    {
+        cppy::ptr owner( PyObject_CallNoArgs( self->atomref ) );
```

Comment:
This need to be check for an exception thrown. After this line, you need:

```
if( !owner)
{ 
    return 0;
}
```

Reply:
It's been too long since I did C Pyrhon API stuff I should have seen this.

Reply:
I did a search and it looks like atomlist misses these as well? https://github.com/nucleic/atom/blob/main/atom/src/atomlist.cpp#L788-L791

Reply:
That slipped by me, but needs to be corrected: @MatthieuDartiailh 
https://github.com/nucleic/atom/blame/main/atom/src/atomlist.cpp#L788-L791